contracts/chain-adapters/l2/Ovm_WithdrawalAdapter.sol

```diff
@@ -0,0 +1,138 @@
+     */
+    constructor(
+        IERC20 _l2Usdc,
+        ITokenMessenger _cctpTokenMessenger,
+        address _l2Gateway,
+        IOvm_SpokePool _spokePool
+    ) WithdrawalAdapter(_l2Usdc, _cctpTokenMessenger, _l2Gateway) {
+        spokePool = _spokePool;
+        wrappedNativeToken = spokePool.wrappedNativeToken();
+        l2Eth = spokePool.l2Eth();
+        l1Gas = spokePool.l1Gas();
+    }
+
+    /*
+     * @notice Calls CCTP or the Optimism token gateway to withdraw tokens back to the recipient.
+     * @param recipient L1 address of the recipient.
+     * @param l2TokenAddress address of the l2Token to send back.
+     * @param amountToReturn amount of l2Token to send back.
+     */
+    function withdrawToken(
+        address recipient,
+        address,
+        address l2TokenAddress,
+        uint256 amountToReturn
+    ) public override {
+        // If the token being bridged is WETH then we need to first unwrap it to ETH and then send ETH over the
+        // canonical bridge. On Optimism, this is address 0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000.
+        if (l2TokenAddress == address(wrappedNativeToken)) {
+            WETH9Interface(l2TokenAddress).withdraw(amountToReturn); // Unwrap into ETH.
+            l2TokenAddress = l2Eth; // Set the l2TokenAddress to ETH.
+            IL2ERC20Bridge(Lib_PredeployAddresses.L2_STANDARD_BRIDGE).withdrawTo{ value: amountToReturn }(
+                l2TokenAddress, // _l2Token. Address of the L2 token to bridge over.
+                recipient, // _to. Withdraw, over the bridge, to the l1 pool contract.
+                amountToReturn, // _amount.
+                l1Gas, // _l1Gas. Unused, but included for potential forward compatibility considerations
+                "" // _data. We don't need to send any data for the bridging action.
+            );
+        }
+        // If the token is USDC && CCTP bridge is enabled, then bridge USDC via CCTP.
+        else if (_isCCTPEnabled() && l2TokenAddress == address(usdcToken)) {
+            _transferUsdc(recipient, amountToReturn);
+        }
+        // Note we'll default to withdrawTo instead of bridgeERC20To unless the remoteL1Tokens mapping is set for
+        // the l2TokenAddress. withdrawTo should be used to bridge back non-native L2 tokens
+        // (i.e. non-native L2 tokens have a canonical L1 token). If we should bridge "native L2" tokens then
+        // we'd need to call bridgeERC20To and give allowance to the tokenBridge to spend l2Token from this contract.
+        // Therefore for native tokens we should set ensure that remoteL1Tokens is set for the l2TokenAddress.
+        else {
+            IL2ERC20Bridge tokenBridge = IL2ERC20Bridge(
+                spokePool.tokenBridges(l2TokenAddress) == address(0)
+                    ? Lib_PredeployAddresses.L2_STANDARD_BRIDGE
+                    : spokePool.tokenBridges(l2TokenAddress)
+            );
+            if (spokePool.remoteL1Tokens(l2TokenAddress) != address(0)) {
```

Comment:
prefer using spokePool.tokenBridges(l2Token) in all cases to make it easier to manage all the token mappings.

Reply:
It was factored out here https://github.com/across-protocol/contracts/pull/605/commits/4db2572d1a4b4bf7ebb81d87659f57e3afda4f0b